### PR TITLE
expand on alternatives that were considered and rejected for sidecars

### PR DESCRIPTION
**Pull request by Joseph-Irving**
Here's my first pass at expanding this section, let me know what you think

### CODE REVIEW

File: keps/sig-apps/0753-sidecarcontainers.md [hunk 1]
```diff
@@ -328,9 +334,82 @@ Older Kubelets should still be able to schedule Pods that have sidecar container
...
+```yaml
+containers:
+  - name: myApp
+    startupOrder: 2
+    shutdownOrder: 1
+  - name: mySidecar
+    startupOrder: 2
+    shutdownOrder: 1
+  ```
+  This was dismissed as the UX was considered overly complex for the use cases we were trying to enable and also lacked the semantics around container shutdown triggering for things like Jobs.
```
Comment (alban) [inline keps/sig-apps/0753-sidecarcontainers.md:415]:
````suggestion
```
This was dismissed as the UX was considered overly complex for the use cases we were trying to enable and also lacked the semantics around container shutdown triggering for things like Jobs.
````

File: keps/sig-apps/0753-sidecarcontainers.md [hunk 2]
```diff
@@ -328,9 +334,82 @@ Older Kubelets should still be able to schedule Pods that have sidecar container
 [stalled]: https://github.com/kubernetes/enhancements/issues/753#issuecomment-597372056
 
 ## Alternatives
+This section contains ideas that were originally discussed but then dismissed in favour of the current design.
+### pod.SidecarContainers
```
Comment (rata) [inline keps/sig-apps/0753-sidecarcontainers.md:338]: Maybe something like this is clearer?
```suggestion
### Add pod.Spec.SidecarContainers array
```

File: keps/sig-apps/0753-sidecarcontainers.md [hunk 3]
```diff
@@ -328,9 +334,82 @@ Older Kubelets should still be able to schedule Pods that have sidecar container
...
+
+### Depends On
+Similar to [systemd](https://www.freedesktop.org/wiki/Software/systemd/) this would allow you to specify that a container depends on another container, preventing that container from starting until the container it depends on has also started. This could also be used in shutdown to ensure that the containers which have dependent containers are only terminated after their dependents have all safely shut down.
+```yaml
+containers:
+  - name: myApp
+    dependsOn: mySidecar
+  - name: mySidecar
+```
+This was rejected as the UX was considered to be overly complicated for the use cases that we were trying to solve. It also doesn't solve the problem of Job termination where you want the sidecars to be terminated once the main containers have exited, to do that you would require some additional fields that would further complicate the UX.
```
Comment (rata) [inline keps/sig-apps/0753-sidecarcontainers.md:393]: I think that, while complex, this idea has merit for future use cases. Maybe we should mention that too?

File: keps/sig-apps/0753-sidecarcontainers.md [hunk 4]
```diff
@@ -328,9 +334,82 @@ Older Kubelets should still be able to schedule Pods that have sidecar container
...
+```yaml
+initContainers:
+  - name: myInit
+containers:
+  - name: myApp
+sidecarContainers:
+  - name: mySidecar
+```
+
+### Primary Container
```
Comment (rata) [inline keps/sig-apps/0753-sidecarcontainers.md:350]: Maybe something like this?
```suggestion
### Mark some containers as "primary" for the pod
```

File: keps/sig-apps/0753-sidecarcontainers.md [hunk 5]
```diff
@@ -328,9 +334,82 @@ Older Kubelets should still be able to schedule Pods that have sidecar container
...
+      - name: worker
+        command: ["do a job"]
+      - name: "sidecar"
+        command: ["help"]
+  backoffLimit: 4
+  primaryContainer: worker
+```
 
-Another alternative would be to change the Job Spec to have a `primaryContainer` field to tell it which containers are important. However I feel this is perhaps too specific to job when this Sidecar concept could be useful in other scenarios.
+### Sidecar: true
```
Comment (rata) [inline keps/sig-apps/0753-sidecarcontainers.md:366]:
```suggestion
### Have a boolean to mark sidecar containers (sidecar: true)
```

File: keps/sig-apps/0753-sidecarcontainers.md [hunk 6]
```diff
@@ -328,9 +334,82 @@ Older Kubelets should still be able to schedule Pods that have sidecar container
...
+```yaml
+containers:
+  - name: myApp
+  - name: mySidecar
+    sidecar: true
+```
+A boolean flag of `sidecar: true` could be used to indicate which pods are sidecars, this was dismissed as it was considered too specific and potentially other types of container lifecycle may want to be added in the future.
 
-A boolean flag of `sidecar: true` could be used to indicate which pods are sidecars, however this prevents additional ContainerLifecycles from being added in the future.
+### terminationFatalToPod: true
```
Comment (rata) [inline keps/sig-apps/0753-sidecarcontainers.md:375]:
```suggestion
### Mark which containers, when terminated, should kill the pod
```

File: keps/sig-apps/0753-sidecarcontainers.md [hunk 7]
```diff
@@ -328,9 +334,82 @@ Older Kubelets should still be able to schedule Pods that have sidecar container
...
+This suggestion was to have the ability to mark certain containers as critical to the pod, if they exited it would cause the other containers to exit. While this helped solve things like Jobs it didn't solve the wider issue of ordering startup and shutdown.
+
+```yaml
+containers:
+  - name: myApp
+    terminationFatalToPod: true
+  - name: mySidecar
+```
+
+### Depends On
```
Comment (rata) [inline keps/sig-apps/0753-sidecarcontainers.md:385]:
```suggestion
### Add "Depends On" semantics to containers
```

File: keps/sig-apps/0753-sidecarcontainers.md [hunk 8]
```diff
@@ -328,9 +334,82 @@ Older Kubelets should still be able to schedule Pods that have sidecar container
...
+Similar to [systemd](https://www.freedesktop.org/wiki/Software/systemd/) this would allow you to specify that a container depends on another container, preventing that container from starting until the container it depends on has also started. This could also be used in shutdown to ensure that the containers which have dependent containers are only terminated after their dependents have all safely shut down.
+```yaml
+containers:
+  - name: myApp
+    dependsOn: mySidecar
+  - name: mySidecar
+```
+This was rejected as the UX was considered to be overly complicated for the use cases that we were trying to solve. It also doesn't solve the problem of Job termination where you want the sidecars to be terminated once the main containers have exited, to do that you would require some additional fields that would further complicate the UX.
+
+### Startup/Shutdown Phases/Numbers
```
Comment (rata) [inline keps/sig-apps/0753-sidecarcontainers.md:395]:
```suggestion
### Use arbitrary numbers to select the startup/shutdown order
```

File: keps/sig-apps/0753-sidecarcontainers.md (hunk 4 above)
Comment (Joseph-Irving) [reply keps/sig-apps/0753-sidecarcontainers.md:350]: This idea was that you specifically chose one container, not multiple

File: keps/sig-apps/0753-sidecarcontainers.md (hunk 3 above)
Comment (Joseph-Irving) [reply keps/sig-apps/0753-sidecarcontainers.md:393]: I don't know if it's worth explicitly calling that out, people can determine that for themselves. The purpose of this section was just to try and capture the things that were rejected with a bit of the rationale for why.

File: keps/sig-apps/0753-sidecarcontainers.md (hunk 8 above)
Comment (Joseph-Irving) [reply keps/sig-apps/0753-sidecarcontainers.md:395]: It wasn't just numbers, we also had the idea of phases which are essentially enums to achieve a similar thing

**Comment by Joseph-Irving** [comment]
I can find some of the comments on github discussing some of these things, however, if you want some kind of record of a definitive decision being made on each of these alternatives that probably doesn't exist. The decision was  made when the KEP approvers from Sig-apps and Sig-node approved the KEP and allowed it to be merged, so there's no record of them rejecting the alternatives only of them accepting the current proposal. Also just because something did get rejected earlier doesn't mean it never gets discussed again. I've had the same conversation with many people about the same things often reaching the same conclusion. Perhaps if we started something like this from the beginning that would've at least helped to reduce endless debates on the same topic.

File: keps/sig-apps/0753-sidecarcontainers.md (hunk 4 above)
Comment (rata) [reply keps/sig-apps/0753-sidecarcontainers.md:350]: Ohh, didn't realize it was restricted only to one (that seems like a tricky interface), but in that case s/some containers/a container/? :)

File: keps/sig-apps/0753-sidecarcontainers.md (hunk 8 above)
Comment (rata) [reply keps/sig-apps/0753-sidecarcontainers.md:395]: Oh, okay. I think something more descriptive might make sense, though. Not sure this is a good name, but maybe something like this, then?

```suggestion
Predefined startup/shutdown phases (ala "run-level" in sysv init systems)
```

**Comment by rata** [comment]
@Joseph-Irving I think those links might help, and maybe we can add what you say here too: We added some links to where *part* of the discussion happened and mention that there are no records of some alternatives and the discussions were repeated arriving to the same conclusions, and we hope that adding this section would at least give some context to the reader and save time if we will arrive to the same conclusion.

What do you think?

File: keps/sig-apps/0753-sidecarcontainers.md (hunk 3 above)
Comment (rata) [reply keps/sig-apps/0753-sidecarcontainers.md:393]: Maybe not here, but I think I'll try to mention it in some part of the KEP. Being open about the possible alternatives that seem more likely and, if possible, why this KEP seems extensible if those use cases are better understood in the future seems to make a more robust KEP, IMHO.

**Comment by Joseph-Irving** [comment]
I added all the links I could find, I may have missed something, there's a lot of discussion spread over many PRs.

File: keps/sig-apps/0753-sidecarcontainers.md (hunk 3 above)
Comment (rata) [reply keps/sig-apps/0753-sidecarcontainers.md:401]: Will take note to see if I can fit this in another part later :)

**Pull request merged**